### PR TITLE
21 uniform buffer definition and creation

### PR DESCRIPTION
**Pull request by mcr431**
This is the first section of uniform buffer tutorial which includes adding the struct to our rust code and the uniform buffer to our shader

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -332,7 +332,26 @@ https://vulkan-tutorial.com/Vertex_buffers/Index_buffer
...
+
+Note that unlike the original tutorial we did **not** need to create any layout binding. This is handled internally by vulkano when creating
+a descriptor set, as we'll see in the next section.
+
+At this point our program will compile and run but immediately panic because we specify a binding in our shader but do not
+include a matching descriptor set. 
+
+[Vertex Shader Diff](src/bin/21_shader_uniformbuffer.vert.diff) / [Vertex Shader](src/bin/21_shader_uniformbuffer.vert)
+
+[Diff](src/bin/21_descriptor_layout_and_buffer.rs) / [Complete code](src/bin/21_descriptor_layout_and_buffer.rs)
```
Comment (bwasty) [inline README.md]: There's a `.diff` missing for the diff link.

File: src/bin/21_descriptor_layout_and_buffer.rs [hunk 2]
```diff
@@ -0,0 +1,727 @@
+#![feature(duration_as_u128)]
```
Comment (bwasty) [inline src/bin/21_descriptor_layout_and_buffer.rs]: ```#![feature] may not be used on the stable release channel```
-> no unstable features please unless there's a very good reason.

File: README.md (hunk 1 above)
Comment (bwasty) [inline README.md]: There's a `.diff` missing for the diff.

File: src/bin/21_descriptor_layout_and_buffer.rs [hunk 3]
```diff
@@ -0,0 +1,727 @@
...
+}
+impl Vertex {
+    fn new(pos: [f32; 2], color: [f32; 3]) -> Self {
+        Self { pos, color }
+    }
+}
+impl_vertex!(Vertex, pos, color);
+
+#[derive(Copy, Clone)]
+#[allow(dead_code)]
```
Comment (bwasty) [inline src/bin/21_descriptor_layout_and_buffer.rs]: That doesn't look necessary (anymore).

File: src/bin/21_descriptor_layout_and_buffer.rs (hunk 2 above)
Comment (mcr431) [reply src/bin/21_descriptor_layout_and_buffer.rs]: Just seeing these comments now. I'll update this to not use features tomorrow.

File: src/bin/21_descriptor_layout_and_buffer.rs (hunk 3 above)
Comment (mcr431) [reply src/bin/21_descriptor_layout_and_buffer.rs]: Will update you when fixed

File: README.md (hunk 1 above)
Comment (bwasty) [reply README.md]: Still links to the .rs, not the .rs.diff.

File: src/bin/21_descriptor_layout_and_buffer.rs [hunk 4]
```diff
@@ -0,0 +1,723 @@
...
+        Self { pos, color }
+    }
+}
+impl_vertex!(Vertex, pos, color);
+
+#[derive(Copy, Clone)]
+struct UniformBufferObject {
+    model: glm::Mat4,
+    view: glm::Mat4,
+    proj: glm::Mat4,
```
Comment (bwasty) [inline src/bin/21_descriptor_layout_and_buffer.rs]: There are a bunch of `field is never used` warnings here and in some other places.
I also noticed there are now quite a few `clippy` warnings in all steps - would be also nice to clean them up eventually (or `allow` them).

Comment (mcr431) [reply src/bin/21_descriptor_layout_and_buffer.rs]: Noted. I'm going to clean everything up today and then test on my mac as I haven't done that yet.

**Comment by bwasty** [comment]
It's a bit annoying that Github still shows the commits from #5 although they're exactly the same as on master...anyway, merging :)

**Pull request merged**